File: src/configuration_manager/config_models.py
"""Configuration models module."""

from dataclasses import dataclass, field


@dataclass
class ModelSelectionConfig:
    model_name: str = "blip2"


@dataclass
class MultiModalConfig:
    blip: str = "Salesforce/blip2-opt-2.7b"
    vit: str = "nlpconnect/vit-gpt2-image-captioning"
    llava: str = "llava-hf/llava-1.5-7b-hf"


@dataclass
class Variants:
    phi3: str = "phi3"
    llama3: str = "llama3"
    gemma2: str = "gemma2"


@dataclass
class OllamaConfig:
    variants: Variants = field(default_factory=Variants)
    use: str = "Phi"
    temperature: int = 1
    top_p: float = 0.9
    stream: bool = False


@dataclass
class ImageCompressionConfig:
    type: str = "webp"
    compress: bool = True
    compression_quality: int = 50
    resize_factor: float = 0.5


@dataclass
class TransformConfig:
    resize: tuple = (256, 256)
    center_crop: int = 224

File: src_cpp/configuration_manager/src/ConfigModels.cpp
#include "ConfigModels.h"
#include <stdexcept>

void AppConfig::validate() {
    // Validate MultiModalConfig
    if (multimodal.blip.empty()) {
        throw std::invalid_argument("The 'blip' field in MultiModalConfig must be a non-empty string.");
    }
    if (multimodal.vit.empty()) {
        throw std::invalid_argument("The 'vit' field in MultiModalConfig must be a non-empty string.");
    }

    // Validate OllamaConfig
    if (ollama.variants.phi3.empty()) {
        throw std::invalid_argument("The 'phi3' field in Variants must be a non-empty string.");
    }
    if (ollama.variants.llama3.empty()) {
        throw std::invalid_argument("The 'llama3' field in Variants must be a non-empty string.");
    }
    if (ollama.variants.gemma2.empty()) {
        throw std::invalid_argument("The 'gemma2' field in Variants must be a non-empty string.");
    }
    if (ollama.use.empty()) {
        throw std::invalid_argument("The 'use' field in OllamaConfig must be a non-empty string.");
    }

    // Validate ImageCompressionConfig
    if (image_compression.type.empty()) {
        throw std::invalid_argument("The 'type' field in ImageCompressionConfig must be a non-empty string.");
    }

    // Validate ModelSelectionConfig
    if (model_selection.model_name.empty()) {
        throw std::invalid_argument("The 'model_name' field in ModelSelectionConfig must be a non-empty string.");
    }

    // Validate ChromaDBConfig
    if (chroma_db.blip.empty()) {
        throw std::invalid_argument("The 'blip' field in ChromaDBConfig must be a non-empty string.");
    }
    if (chroma_db.llava.empty()) {
        throw std::invalid_argument("The 'llava' field in ChromaDBConfig must be a non-empty string.");
    }
}
